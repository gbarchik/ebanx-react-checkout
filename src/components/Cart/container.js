import { connect } from 'react-redux'

import { loadProducts } from '../../store/reducers/products'
import { addToCart, removeFromCart } from '../../store/reducers/cart'
import { showForm, hideForm } from '../../store/reducers/payment'

import Cart from './index'

const mapDispatchToProps = {
    loadProducts,
    addToCart,
    removeFromCart,
    showForm,
    hideForm
}

const mapStateToProps = (state) => {
    const { products } = state.products
    const { cart } = state.cart

    return ({
        products,
        cart
    })
}

export default connect(mapStateToProps, mapDispatchToProps)(Cart)