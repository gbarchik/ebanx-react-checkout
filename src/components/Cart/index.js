import React, { Component } from 'react'
import PropTypes from 'prop-types'

import GridContainer from '../GridContainer'
import ProductsList from '../ProductsList'
import ProductRow from '../ProductRow'
import CartTotal from '../CartTotal'
import Spacer from '../Spacer'

class Cart extends Component {
    componentDidMount() {
        // Load all the products
        this.props.loadProducts();
    }

    componentDidUpdate() {
        // If there is any products in the card,
        // show the checkout form
        if (this.props.cart.products.length) {
            this.props.showForm()
        } else {
            this.props.hideForm()
        }
    }

    render() {
        const {
            products,
            addToCart,
            removeFromCart,
            cart
        } = this.props;

        return (
            <GridContainer>
                <ProductsList products={products}>
                    {products.map((product) => (
                        <ProductRow
                            key={product.id}
                            product={product}
                            cart={cart}
                            addToCart={() => addToCart(product)}
                            removeFromCart={() => removeFromCart(product)} />
                    ))}
                </ProductsList>
                <Spacer />
                <CartTotal cart={cart} />
            </GridContainer>
        )
    }
}

Cart.propTypes = {
    loadProducts: PropTypes.func.isRequired,
    products: PropTypes.array.isRequired,
    showForm: PropTypes.func.isRequired,
    hideForm: PropTypes.func.isRequired,
    addToCart: PropTypes.func.isRequired,
    removeFromCart: PropTypes.func.isRequired,
    cart: PropTypes.object.isRequired
}

export default Cart