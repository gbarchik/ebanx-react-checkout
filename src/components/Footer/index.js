import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import GridContainer from '../GridContainer'
import Spacer from '../Spacer'

import {
    globalPadding,
    headerFontFamily,
    headerFontWeight
} from '../../helpers/foundation'
import { calculateRem } from '../../helpers/helper'

const StyledFooter = styled.footer`
    padding: ${calculateRem(4 * globalPadding)} 0 ${calculateRem(globalPadding)}
`

const FooterTitle = styled.span`
    display: block;
    margin: 0;
    font-family: ${headerFontFamily};
    font-weight: ${headerFontWeight};
`

const Footer = (props) => {
    const { title, poweredBy } = props.config

    return (
        <StyledFooter>
            <GridContainer>
                <Spacer />
                <FooterTitle>{title}</FooterTitle>
                {poweredBy && <span>Powered by {poweredBy}</span>}
            </GridContainer>
        </StyledFooter>
    )
}

Footer.propType = {
    config: PropTypes.shape({
        title: PropTypes.string.isRequired,
        poweredBy: PropTypes.string
    })
}

export default Footer
