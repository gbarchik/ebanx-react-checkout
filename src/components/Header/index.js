import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import GridContainer from '../GridContainer'

import {
    globalPadding,
    globalMargin,
    headerFontFamily,
    headerFontSizes,
    headerFontWeight
} from '../../helpers/foundation'
import { calculateRem, calculateEm } from '../../helpers/helper'

const StyledHeader = styled.header`
    padding: ${calculateRem(4 * globalPadding)} 0 ${calculateRem(globalPadding)}
`

const HeaderTitle = styled.h1`
    margin: 0 0 ${calculateEm((1.5 * globalMargin), headerFontSizes.h1)};
    font-family: ${headerFontFamily};
    font-size: ${calculateRem(headerFontSizes.h1)};
    font-weight: ${headerFontWeight};
`

const Header = (props) => {
    const { title, description } = props.config

    return (
        <StyledHeader>
            <GridContainer>
                <HeaderTitle>{title}</HeaderTitle>
                {description && <p>{description}</p>}
            </GridContainer>
        </StyledHeader>
    )
}
    

Header.propType = {
    config: PropTypes.shape({
        title: PropTypes.string.isRequired,
        description: PropTypes.string
    })
}

export default Header
