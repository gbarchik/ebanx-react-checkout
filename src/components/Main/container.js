import { connect } from 'react-redux'

import Main from './index'

const mapStateToProps = (state) => {
    const { checkout } = state

    return ({
        checkout
    })
}

export default connect(mapStateToProps)(Main)