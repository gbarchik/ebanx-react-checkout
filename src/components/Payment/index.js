import React, { Component } from 'react'
import PropTypes from 'prop-types'

import GridContainer from '../GridContainer'
import Grid from '../Grid'
import GridColumn from '../GridColumn'
import PaymentStep from '../PaymentStep'
import Label from '../Form/Label'
import ErrorMessage from '../Form/ErrorMessage'
import Input from '../Form/Input'
import Select from '../Form/Select'
import PaymentSelect from '../Form/PaymentSelect'
import Button from '../Form/Button'

// Helpers to check form validation
const validateStep = (step) => step.fields.reduce((isValid, field) => {
    if (field.conditional && !showConditional(field.conditional, step)) {
        return isValid
    }

    return field.valid.isValid ? isValid : false
}, true)

const validateInput = (val, validators) =>
    validators.reduce((valid, validator) =>
        (valid.isValid && !validator.validator(val)) ? { isValid: false, message: validator.message } : valid
        , { isValid: true, message: '' })

// Helper for conditional input
const showConditional = (conditional, step) => {
    let shouldShow = true

    if (conditional) {
        const conditionalInput = step.fields.filter((field) => field.name === conditional.inputName)
        if (conditionalInput[0].val !== conditional.inputValue) {
            shouldShow = false
        }
    }

    return shouldShow
}


class Payment extends Component {
    constructor(props) {
        super(props)

        this.handleInputChange = this.handleInputChange.bind(this)
        this.handleInputBlur = this.handleInputBlur.bind(this)
    }

    handleInputChange(e, fieldIndex, step, stepIndex) {
        const val = e.target.value
        let newStep = {
            ...step,
            fields: step.fields.map((field, i) => {
                return fieldIndex === i ? {
                    ...field,
                    val,
                    valid: validateInput(val, field.validators)
                } : {
                    ...field
                }
            })
        }

        // Validates the step
        newStep.isComplete = validateStep(newStep)

        // Updates the value of the field,
        // checks its validity and the
        // step validity
        this.props.updateStep(newStep, stepIndex)

        // If the step is valid, allow the user to fill the next step
        // If a previous step is invalid, return to that step
        if ((stepIndex === this.props.payment.step) && newStep.isComplete) {
            this.props.goToStep(stepIndex + 1)
        } else if ((stepIndex < this.props.payment.step) && !newStep.isComplete) {
            this.props.goToStep(stepIndex)
        }
    }

    handleInputBlur(e, fieldIndex, step, stepIndex) {

        // Updates the field dirty state
        this.props.updateStep({
            ...step,
            fields: step.fields.map((field, i) => {
                return fieldIndex === i ? {
                    ...field,
                    dirty: e.target.value ? true : field.dirty
                } : {
                    ...field
                }
            })
        }, stepIndex)
    }

    render() {
        const { payment, cart, buyNow } = this.props
        const { steps } = payment

        const getField = (field, fieldIndex, step, stepIndex) => {
            const {
                size,
                name,
                title,
                type,
                placeholder,
                val,
                options,
                mask,
                dirty,
                conditional,
                valid
            } = field

            let TheField = ''

            switch (type) {
                case 'select':
                    TheField = (
                        <Select
                            dirty={dirty}
                            placeholder={placeholder}
                            id={name}
                            value={val}
                            options={options}
                            onChange={(e) => this.handleInputChange(e, fieldIndex, step, stepIndex)}
                            onBlur={(e) => this.handleInputBlur(e, fieldIndex, step, stepIndex)}
                        />
                    )
                    break

                case 'paymentSelect':
                    TheField = (
                        <PaymentSelect
                            id={name}
                            value={val}
                            options={options}
                            onChange={(e) => this.handleInputChange(e, fieldIndex, step, stepIndex)}
                            onBlur={(e) => this.handleInputBlur(e, fieldIndex, step, stepIndex)}
                        />
                    )
                    break

                default:
                    TheField = (
                        <Input
                            placeholder={placeholder}
                            id={name}
                            type={type}
                            mask={mask}
                            value={val}
                            onChange={(e) => this.handleInputChange(e, fieldIndex, step, stepIndex)}
                            onBlur={(e) => this.handleInputBlur(e, fieldIndex, step, stepIndex)}
                        />
                    )
            }

            return (
                <GridColumn medium={size} key={name} style={showConditional(conditional, step) ? {} : { display: 'none' }}>
                    <Label htmlFor={name}>{title}</Label>
                    {TheField}
                    {(dirty && !valid.isValid) && <ErrorMessage>{valid.message}</ErrorMessage>}
                </GridColumn>
            )
        }

        return (
            <GridContainer>
                {steps && steps.map((step, stepIndex) => (
                    <PaymentStep
                        key={step.title}
                        title={step.title}
                        stepNumber={(stepIndex + 1)}
                        contentVisible={step.isVisible && payment.isFormVisible}>

                        <Grid>
                            {step.fields && step.fields.map((field, fieldIndex) => {
                                return getField(field, fieldIndex, step, stepIndex)
                            })}

                            {stepIndex === 2 && (
                                <GridColumn>
                                    <Button
                                        type={'button'}
                                        disabled={payment.step !== 3}
                                        onClick={() => buyNow(cart, steps)}
                                    >BUY NOW</Button>
                                </GridColumn>
                            )}
                        </Grid>
                    </PaymentStep>
                ))}
            </GridContainer>
        )
    }
}

Payment.propTypes = {
    updateStep: PropTypes.func.isRequired,
    goToStep: PropTypes.func.isRequired,
    buyNow: PropTypes.func.isRequired,
    payment: PropTypes.shape({
        steps: PropTypes.array.isRequired
    }),
    cart: PropTypes.object.isRequired
}

export default Payment