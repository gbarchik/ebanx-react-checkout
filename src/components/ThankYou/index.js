import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import GridContainer from '../GridContainer'

import {
    colors,
    globalPadding,
    headerFontFamily,
    headerFontSizes,
    headerFontWeight
} from '../../helpers/foundation'
import { calculateRem, calculateEm } from '../../helpers/helper'

const StyledThankYou = styled.div`
    padding: ${calculateEm(5 * globalPadding)} ${calculateEm(2 * globalPadding)}
    text-align: center;
    background-color: ${colors.lightGray}
`

const ThankYouTitle = styled.h2`
    position: relative;
    display: inline-block;
    padding-left: ${calculateEm(45, headerFontSizes.h2)};
    font-family: ${headerFontFamily};
    font-size: ${calculateRem(headerFontSizes.h2)};
    font-weight: ${headerFontWeight};

    &:after {
        content: '';
        position: absolute;
        top: 50%;
        left: 0;
        display: block;
        width: ${calculateEm(35, headerFontSizes.h2)};
        height: ${calculateEm(35, headerFontSizes.h2)};
        border-radius: 100%;
        background-color: ${colors.mediumGray};
        transform: translateY(-50%);
    }
`

const ThankYou = (props) => {
    const { title, message } = props.checkout

    return (
        <GridContainer>
            <StyledThankYou>
                <ThankYouTitle>{title}</ThankYouTitle>
                <p dangerouslySetInnerHTML={{ __html: message }} />
            </StyledThankYou>
        </GridContainer>
    )
}

ThankYou.propTypes = {
    checkout: PropTypes.shape({
        title: PropTypes.string.isRequired,
        message: PropTypes.string.isRequired
    })
}

export default ThankYou