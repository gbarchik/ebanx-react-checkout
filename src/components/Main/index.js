import React from 'react'

import Header from '../Header/container'
import Cart from '../Cart/container'
import Payment from '../Payment/container'
import Footer from '../Footer/container'
import ThankYou from '../ThankYou/container'

const Main = (props) => {
    const { bought } = props.checkout

    return (
        <div>
            <Header />
            {bought ? (
                <ThankYou />
            ) : (
                <div>
                    <Cart />
                    <Payment />
                </div>
            )}
            <Footer />
        </div>
    )

}

export default Main