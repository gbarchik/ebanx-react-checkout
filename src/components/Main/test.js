import React from 'react'
import { shallow } from 'enzyme'

import Main from './index'
import Header from '../Header/container'
import Footer from '../Footer/container'
import Cart from '../Cart/container'
import Payment from '../Payment/container'
import ThankYou from '../ThankYou/container'

const fakeProps = {
    checkout: {
        bought: false
    }
}

describe('(Layout) Main', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<Main {...fakeProps} />)
        expect(wrapper).toMatchSnapshot()
    })

    it('renders as a <div>', () => {
        const wrapper = shallow(<Main {...fakeProps} />)
        expect(wrapper.type()).toBe('div')
    })

    it('should render a <Header>', () => {
        const wrapper = shallow(<Main {...fakeProps} />)
        expect(wrapper.find(Header)).toHaveLength(1)
    })

    it('should render a <Footer>', () => {
        const wrapper = shallow(<Main {...fakeProps} />)
        expect(wrapper.find(Footer)).toHaveLength(1)
    })

    it('should render a <Cart> and a <Payment> if bought is false', () => {
        const wrapper = shallow(<Main {...fakeProps} />)
        wrapper.setProps({ checkout: { bought: false } })
        expect(wrapper.find(Cart)).toHaveLength(1)
        expect(wrapper.find(Payment)).toHaveLength(1)
    })

    it('should not render a <ThankYou> if bought is false', () => {
        const wrapper = shallow(<Main {...fakeProps} />)
        wrapper.setProps({ checkout: { bought: false } })
        expect(wrapper.find(ThankYou)).toHaveLength(0)
    })

    it('should not render a <Cart> and a <Payment> if bought is true', () => {
        const wrapper = shallow(<Main {...fakeProps} />)
        wrapper.setProps({ checkout: { bought: true } })
        expect(wrapper.find(Cart)).toHaveLength(0)
        expect(wrapper.find(Payment)).toHaveLength(0)
    })

    it('should render a <ThankYou> if bought is true', () => {
        const wrapper = shallow(<Main {...fakeProps} />)
        wrapper.setProps({ checkout: { bought: true } })
        expect(wrapper.find(ThankYou)).toHaveLength(1)
    })
})
