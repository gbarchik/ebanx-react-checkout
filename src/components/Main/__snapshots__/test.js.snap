// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Layout) Main renders correctly 1`] = `
<div>
  <Connect(Header) />
  <div>
    <Connect(Cart) />
    <Connect(Payment) />
  </div>
  <Connect(Footer) />
</div>
`;
